### ISSUE TITLE
Remove boost::variant from neural network code

### ISSUE THREAD

**Issue by zoq** [mlpack/mlpack#2779]
This is a follow-up for https://github.com/mlpack/mlpack/issues/2647. Based on the experiments in https://github.com/mlpack/mlpack/pull/2777 we figured it makes sense to remove the boost::variant approach. This should not only make it easier for new-comers to contribute to the network codebase (the boost::variant and visitor approach can be confusing) but also reduce the memory usage during the build (templates tend to consume memory). Anyway, there are various layers we have to update (a list of layers and steps can be found in #2777). This is a good first issue to get familiar with the network codebase.

**Comment by kaushal07wick** [comment]
hey @zoq ! Is this issue still open I would like to fix this problem. Thanks

**Comment by zoq** [comment]
@kaushal07wick yes, see https://github.com/mlpack/mlpack/pull/2777 for a list of layers that aren't taken up.

**Comment by kaushal07wick** [comment]
Ok thanks I would look onto it.

**Comment by rajpratyush** [comment]
Hi @zoq I would like to know if this issue is still open for work or not.
In case it is I would like to work on it

**Comment by shubhamabhang77** [comment]
Is there anyone working on this issue? If not I would like to take this.

**Comment by karakotineeraj** [comment]
Hi @zoq I would like to work on this issue and if you could, can you tell me where should I look for this in the codebase?

**Comment by rcurtin** [comment]
I've been going through the `ann-vtable` branch to refactor every layer.  It's going to be a little while until I manage to actually make things work correctly (surely I can make some intermediate PRs, but we won't be able to merge them here for a while...).  I wanted to write down a few observations I had for discussion, in case anyone has alternative perspectives:

### 1. Storage of intermediate values in layers

Currently, we store matrices like `outputParameter`, `delta`, `gradient` in each layer.  These are used to store the output, delta, and gradient of each layer, and this provides a convenient interface for `boost::visitor` to work with.  However, these members make the layer definitions confusing, especially since some layers have these parameters and others don't, depending on the layer.

Instead, we can handle each of these things in the enclosing `FFN`/`RNN`/`GAN` class.  I'll quickly describe how we can do this to replace the `outputParameter` member in the forward pass.  First, before actually calling `Forward()` on any layer, we initialize a matrix that will be able to hold the combined outputs of all layers in one contiguous block.  We can do this assuming that we can ask each layer what the size of its output is.  Then, supposing the first layer outputs the matrix `A`, and the second `B`, and so forth, then the matrix we initialize will store `[A; B; ...]`.  Then, we simply need to pass in aliases to each layer's `Forward()`.

Overall, this moves all the logic of the memory management for the forward, backward, and gradient passes to the enclosing network.

### 2. Deterministic parameter (training vs. inference)

Many layers have a `deterministic` parameter, but we can just store this in the base layer `Layer<InputType, OutputType>` and use inheritance to access it.  The enclosing network can pass through each layer and set `deterministic` to `true` when doing inference (or `false` otherwise).

I also think it might be a good idea to change this parameter's name to the more direct name of `training` or something like this.  At least to me, "deterministic" isn't immediately connected to a meaning for me.

### 3. Input shape storage

Lots of layers have `inputWidth` and `inputHeight` and `outputWidth` and `outputHeight`, to handle 2-dimensional inputs.  Our `FFN` class doesn't really support this well, since Armadillo doesn't support higher-dimensional tensors.  So, instead, what we are doing to support this involves passing a matrix of size `n_rows` x `batch_size`, and that `n_rows` is a vectorized/flattened representation of whatever the input object is.

Thus, my idea here is that the base layer has a member `std::vector<size_t> inputDimensions`.  This `inputDimensions` member allows us to represent arbitrary-dimensional tensors as input; so, instead of needing to store `inputWidth` and `inputHeight`, a layer can simply access `inputDimensions[0]` and `inputDimensions[1]`.  In addition, `inputDimensions[2:end]` can be used to represent higher-dimensional objects.

Each layer should implement a function `const std::vector<size_t>& OutputDimensions()`, which can compute the output dimensions assuming that `inputDimensions` is set right.

From the level of `FFN`, when we start a forward pass with given input, we set `inputDimensions` in the first layer, and then proceed through the network to compute `OutputDimensions()` for each layer and use these to set `inputDimensions` for the next layer.  Then, when each layer's `Forward()` is called, we can assume that the input dimensions are correct.

### 4. "Multi-layers", and the 'model' parameters

Some layers hold many layers in them.  I think we can have a base layer `MultiLayer<InputType, OutputType> : Layer<InputType, OutputType>` that internally holds a network and handles `Forward()`/`Backward()`/`Gradient()` for each of those layers.  Of course, each multi-layer can overload those methods if there is some special way that the passes need to be done.  In any case, the multi-layer base class can handle the memory allocations for the things mentioned in the first subsection I wrote above.

If the interface to `MultiLayer<>` is right, this also means the `model` parameter is not necessary anymore and can be removed too.

### 5. Contiguous weight storage

This isn't a change from the existing strategy, but the flow is a little bit different now given the changes above.  When we want to initialize the weights for each layer, we have to follow this process:

1. Take the given input size and push it to the first layer as described in section 3 above, so that `inputDimensions` are computed right for each layer.
2. Pass through each layer and call `WeightSize()` (which can use `inputDimensions`) to get the total number of weights in the network.
3. Allocate a weight matrix that has number of elements equal to the sum of `WeightSize()`s of each layer.
4. Pass a pointer to the right part of the weight matrix to each layer via its `SetWeights(eT*)` method.  Each layer, if it has weights, must wrap the given pointer as an alias.

-----

Anyway, these are very significant changes and will require some significant work.  Once I have some basic tests out of the test suite working, I'll open a PR against `zoq:ann-vtable`, and we can figure out what the way forward is. :+1:

**Comment by zoq** [comment]
> ### 1. Storage of intermediate values in layers
> Currently, we store matrices like `outputParameter`, `delta`, `gradient` in each layer. These are used to store the output, delta, and gradient of each layer, and this provides a convenient interface for `boost::visitor` to work with. However, these members make the layer definitions confusing, especially since some layers have these parameters and others don't, depending on the layer.
> 
> Instead, we can handle each of these things in the enclosing `FFN`/`RNN`/`GAN` class. I'll quickly describe how we can do this to replace the `outputParameter` member in the forward pass. First, before actually calling `Forward()` on any layer, we initialize a matrix that will be able to hold the combined outputs of all layers in one contiguous block. We can do this assuming that we can ask each layer what the size of its output is. Then, supposing the first layer outputs the matrix `A`, and the second `B`, and so forth, then the matrix we initialize will store `[A; B; ...]`. Then, we simply need to pass in aliases to each layer's `Forward()`.
> 
> Overall, this moves all the logic of the memory management for the forward, backward, and gradient passes to the enclosing network.

That basically means you do a dry run over the network to figure out what the output size is, and use that information to initialize e.g. the block for the output matrix?

 I think it doesn't matter if the network branches off (e.g. inception) as long as we asign the correct alias. So somewhat similar to what we currently do with the weights and the `Reset` that is part of the layer implementation.


> ### 2. Deterministic parameter (training vs. inference)
> Many layers have a `deterministic` parameter, but we can just store this in the base layer `Layer<InputType, OutputType>` and use inheritance to access it. The enclosing network can pass through each layer and set `deterministic` to `true` when doing inference (or `false` otherwise).
> 
> I also think it might be a good idea to change this parameter's name to the more direct name of `training` or something like this. At least to me, "deterministic" isn't immediately connected to a meaning for me.

Agreed, I think having the base layer in place allows us to simplify the layers quite a bit. Also looking at the template interface of the BaseLayer. Wondering if we could get away in removing the `InputType` and `OutputType` from the layer and templatize the internal layer functions.

I guess,  another option we have at least for the network weights is to adapt ensmallen, actually I think in some way we already support something that could work, if we interpret the each layer as an independent function, the multi-objective interface would support that.

**Comment by rcurtin** [comment]
> That basically means you do a dry run over the network to figure out what the output size is, and use that information to initialize e.g. the block for the output matrix?

Essentially yes, but it's not like a forward pass is being taken---instead, we are just looping over each of the layers, asking each layer what the output size is (via `OutputDimensions()`), and then collecting the size of the result.  It's quite similar to some of the visitors like WeightSizeVisitor, it's just implemented using inheritance, not the visitor paradigm.

> I think it doesn't matter if the network branches off (e.g. inception) as long as we asign the correct alias.

Actually this might be a bit of an issue.  I haven't thought about network structures like that in my refactoring.  I don't think it's *too* hard to handle though.  Do you have any examples of networks where the layers split and then rejoin later?  I wonder if a better structure than a `std::vector<LayerType*>` might actually be something more general like a DAG then...

> Wondering if we could get away in removing the InputType and OutputType from the layer and templatize the internal layer functions.

I had hoped this too, but the standard does not allow mixing virtual member functions and templates: https://stackoverflow.com/questions/2354210/can-a-class-member-function-template-be-virtual

> I guess, another option we have at least for the network weights is to adapt ensmallen, actually I think in some way we already support something that could work, if we interpret the each layer as an independent function, the multi-objective interface would support that.

Yeah, either that or a "shim" class could work.  I could also imagine some layer-wise optimizer that updates the weights for each layer sequentially or something like this.  One of the things I'd really like to be able to do---eventually---is to have some layers with different weight types.  However, ensmallen's interface currently means that all weights must be in a contiguous matrix; this in turn means that `InputType` and `OutputType` must be the same for every layer in the network.

Technically, this means that for now it would be reasonable to combine `InputType` and `OutputType` into just one parameter---but personally, I would rather keep them separate, in the hopes that someday we will be able to use both of them.  Let me know what you think.

**Comment by rcurtin** [comment]
I finished implementing these suggestions for some basic cases here: https://github.com/zoq/mlpack/pull/2.

**Comment by rcurtin** [comment]
With #2777 merged, I think this is done! :tada:
